apollo-router/src/services/layers/query_analysis.rs

```diff
@@ -0,0 +1,150 @@
+    }
+
+    pub(crate) async fn supergraph_request(
+        &self,
+        request: SupergraphRequest,
+    ) -> Result<SupergraphRequest, SupergraphResponse> {
+        let query = request.supergraph_request.body().query.as_ref();
+
+        if query.is_none() || query.unwrap().trim().is_empty() {
+            let errors = vec![crate::error::Error::builder()
+                .message("Must provide query string.".to_string())
+                .extension_code("MISSING_QUERY_STRING")
+                .build()];
+            tracing::error!(
+                monotonic_counter.apollo_router_http_requests_total = 1u64,
+                status = %StatusCode::BAD_REQUEST.as_u16(),
+                error = "Must provide query string",
+                "Must provide query string"
+            );
+
+            return Err(SupergraphResponse::builder()
+                .errors(errors)
+                .status_code(StatusCode::BAD_REQUEST)
+                .context(request.context)
+                .build()
+                .expect("response is valid"));
+        }
+
+        let op_name = request.supergraph_request.body().operation_name.clone();
+        let query = request
+            .supergraph_request
+            .body()
+            .query
+            .clone()
+            .expect("query presence was already checked");
+        let entry = self
+            .cache
+            .lock()
+            .await
+            .get(&QueryAnalysisKey {
+                query: query.clone(),
+                operation_name: op_name.clone(),
+            })
+            .cloned();
+
+        let compiler = match entry {
+            None => {
+                let (compiler, _) = self.make_compiler(&query);
+
+                let compiler = Arc::new(Mutex::new(compiler));
+
+                (*self.cache.lock().await).put(
+                    QueryAnalysisKey {
+                        query,
+                        operation_name: op_name.clone(),
+                    },
+                    compiler.clone(),
+                );
+
+                compiler
+            }
+            Some(c) => c,
+        };
+
+        let op = match op_name {
+            None => compiler
+                .lock()
+                .await
+                .db
+                .all_operations()
+                .iter()
+                .filter_map(|operation| Operation::from_hir(operation, &self.schema).ok())
+                .next(),
+            Some(name) => compiler
+                .lock()
+                .await
+                .db
+                .all_operations()
+                .iter()
+                .filter_map(|operation| Operation::from_hir(operation, &self.schema).ok())
+                .find(|op| {
+                    if let Some(op_name) = op.name.as_deref() {
+                        op_name == name
+                    } else {
+                        false
+                    }
+                }),
+        };
```

Comment:
This logic should likely be replaced by `compiler.db.find_operation(op_name)` which implements https://spec.graphql.org/October2021/#GetOperation()

Reply:
awesome, this makes the code much nicer.  Fixed in [d579c33](https://github.com/apollographql/router/pull/3200/commits/d579c334219dc7d9b8c46c9e39935f0e08dea02a)